nodejs/dist/uws.js

```diff
@@ -1,98 +1,132 @@
-/* ws interface for Node.js developers */
+'use strict';
 
-const NativeServer = require('./uws_' + process.platform + '_' + process.versions.modules).Server;
 const EventEmitter = require('events');
-const util = require('util');
-
-function Socket(nativeSocket, server) {
-    EventEmitter.call(this);
-    var self = this;
-    this.nativeSocket = nativeSocket;
-    this.server = server;
-
-    this.send = function(message, options) {
-        if (self.nativeSocket !== null) {
-            var binary = false;
-            if (options !== undefined) {
-                binary = options.binary;
-            }
-            self.server.nativeServer.send(self.nativeSocket, message, binary);
-        } else {
-            /* ignore sends on closed sockets */
-        }
+
+const uws = require(`./uws_${process.platform}_${process.versions.modules}`);
+const NativeServer = uws.Server;
+
+class Socket extends EventEmitter {
+    /**
+     * Creates a Socket instance.
+     *
+     * @param {Object} nativeSocket Socket instance
+     * @param {Server} server Server instance
+     */
+    constructor(nativeSocket, server) {
+        super();
+        this.nativeSocket = nativeSocket;
+        this.server = server;
     }
 
-    this.close = function() {
-        if (self.nativeSocket !== null) {
-            self.server.nativeServer.close(self.nativeSocket);
-            self.nativeSocket = null;
-        } else {
-            /* ignore close on closed sockets */
-        }
+    /**
+     * Sends a message.
+     *
+     * @param {String|Buffer} message The message to send
+     * @param {Object} options Send options
+     * @public
+     */
+    send(message, options) {
+        /* ignore sends on closed sockets */
+        if (!this.nativeSocket) return;
+
+        const binary = options && options.binary || false;
+        this.server.nativeServer.send(this.nativeSocket, message, binary);
+    }
+
+    /**
+     * Closes the socket.
+     *
+     * @public
+     */
+    close() {
+        /* ignore close on closed sockets */
+        if (!this.nativeSocket) return;
+
+        this.server.nativeServer.close(this.nativeSocket);
+        this.server = this.nativeSocket = undefined;
```

Comment:
This is causing problems because the close is closing the sockets in the future, so when disconnection callback happens the server needs to be there still.